## Project Description
Sports clothing and athleisure attire is a huge industry, worth approximately <a href="https://www.statista.com/statistics/254489/total-revenue-of-the-global-sports-apparel-market/">$193 billion in 2021</a> with a strong growth forecast over the next decade!

In this notebook, I perform data analysis on the database of an online sports clothing company. I dive into product data such as pricing, reviews, descriptions, and ratings, as well as revenue and website traffic, to improve revenue.

## Data definition

The database <code>sports</code>, contains five tables, with <code>product_id</code> being the primary key for all of them.

<h5 id="info"><code>info</code></h5>
<table>
<thead>
<tr>
<th>columns</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_name</code></td>
<td><code>varchar</code></td>
<td>Name of the product</td>
</tr>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>description</code></td>
<td><code>varchar</code></td>
<td>Description of the product</td>
</tr>
</tbody>
</table>
<h5 id="finance"><code>finance</code></h5>
<table>
<thead>
<tr>
<th>columns</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>listing_price</code></td>
<td><code>float</code></td>
<td>Listing price for product</td>
</tr>
<tr>
<td><code>sale_price</code></td>
<td><code>float</code></td>
<td>Price of the product when on sale</td>
</tr>
<tr>
<td><code>discount</code></td>
<td><code>float</code></td>
<td>Discount, as a decimal, applied to the sale price</td>
</tr>
<tr>
<td><code>revenue</code></td>
<td><code>float</code></td>
<td>Amount of revenue generated by each product, in US dollars</td>
</tr>
</tbody>
</table>
<h5 id="reviews"><code>reviews</code></h5>
<table>
<thead>
<tr>
<th>columns</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_name</code></td>
<td><code>varchar</code></td>
<td>Name of the product</td>
</tr>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>rating</code></td>
<td><code>float</code></td>
<td>Product rating, scored from 1.0 to 5.0</td>
</tr>
<tr>
<td><code>reviews</code></td>
<td><code>float</code></td>
<td>Number of reviews for the product</td>
</tr>
</tbody>
</table>
<h5 id="traffic"><code>traffic</code></h5>
<table>
<thead>
<tr>
<th>columns</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>last_visited</code></td>
<td><code>timestamp</code></td>
<td>Date and time the product was last viewed on the website</td>
</tr>
</tbody>
</table>
<h5 id="brands"><code>brands</code></h5>
<table>
<thead>
<tr>
<th>columns</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>brand</code></td>
<td><code>varchar</code></td>
<td>Brand of the product</td>
</tr>
</tbody>
</table>

### Initializing connection with the database

In [110]:
# Loading sql extension using ipython-sql
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [111]:
# Initializing sql user, host, and database
%sql postgresql://postgres:Postgresql_pr0@localhost/sports

### Counting missing values

In [112]:
%%sql

-- Number of non-null records
SELECT
    COUNT(i.product_id) AS total_rows
    , COUNT(f.listing_price) AS count_list_price
    , COUNT(f.discount) AS count_discount
    , COUNT(f.revenue) AS count_revenue
    , COUNT(t.last_visited) AS count_last_visit
    , COUNT(r.rating) AS count_ratings
    , COUNT(r.reviews) AS count_reviews
FROM
    info AS i
  JOIN
    finance AS f
  USING(product_id)
  JOIN
    traffic AS t
  USING(product_id)
  JOIN
    reviews AS r
  USING(product_id);

 * postgresql://postgres:***@localhost/sports
1 rows affected.


[(3179, 3120, 3120, 3120, 2928, 3120, 3120)]

We can see the database contains 3,179 products in total but most of the important data points are missing for 59 products (3179-3120). Moreover, <code>last_visited</code> is missing around 8% of its values.

Now let's turn our attention to pricing.

### Brands - Nike vs Adidas pricing

How do the price points of Nike and Adidas products differ? This can help us analyse company's stock range and customer market.

Let's assign labels to different price ranges, grouping by <code>brand</code> and <code>price_category</code>, also including the total <code>revenue</code> for each price range and brand.

In [113]:
%%sql

-- Exploring the revenue generated by each brand in different price ranges
SELECT
    b.brand,
    CASE
        WHEN f.listing_price::INTEGER < 42 THEN 'Budget'
        WHEN 42 <= f.listing_price::INTEGER
                AND f.listing_price::INTEGER < 74 THEN 'Average'
        WHEN 74 <= f.listing_price::INTEGER
                AND f.listing_price::INTEGER < 129 THEN 'Expensive'
        ELSE 'Elite'
    END AS price_category
    , COUNT(f.product_id)
    , ROUND(SUM(f.revenue::NUMERIC), 2) AS total_revenue
FROM
    brands AS b
  INNER JOIN
    finance AS f
  USING(product_id)
WHERE
    b.brand IS NOT NULL
    
GROUP BY
    b.brand,
    price_category
ORDER BY
    total_revenue DESC;

 * postgresql://postgres:***@localhost/sports
8 rows affected.


[('Adidas', 'Expensive', 849, Decimal('4626980.07')),
 ('Adidas', 'Average', 1060, Decimal('3233661.06')),
 ('Adidas', 'Elite', 307, Decimal('3014316.83')),
 ('Adidas', 'Budget', 359, Decimal('651661.12')),
 ('Nike', 'Budget', 357, Decimal('595341.02')),
 ('Nike', 'Elite', 82, Decimal('128475.59')),
 ('Nike', 'Expensive', 90, Decimal('71843.15')),
 ('Nike', 'Average', 16, Decimal('6623.50'))]

Grouping products by brand and price range allows us to see that products from **Adidas generate more total revenue regardless of price category**.

Specifically, Adidas products from <code>Elite</code> category priced \$129 or more typically generate the highest revenue, so the company can potentially increase revenue by **shifting their stock to have a larger proportion of Elite products!**

### Average discount by brand

We have been looking at <code>listing_price</code> so far. To understand <code>revenue</code> better, let's take a look at the <code>discount</code>, which is the percent reduction in the <code>listing_price</code> when the product is sold.

We would like to know whether there is a difference in the amount of <code>discount</code> offered between brands, as this could be influencing <code>revenue</code>.

In [114]:
%%sql
 
-- Exploring discounts offered by each brand on average
SELECT
    b.brand
    , ROUND(AVG(f.discount::NUMERIC)*100, 2) AS average_discount
FROM
    brands AS b
  INNER JOIN
    finance AS f
  USING(product_id)
WHERE
    b.brand IS NOT NULL
GROUP BY
    b.brand;

 * postgresql://postgres:***@localhost/sports
2 rows affected.


[('Adidas', Decimal('33.45')), ('Nike', Decimal('0.00'))]

Strangely, **no <code>discount</code>** is offered on Nike products. In comparison, not only do Adidas products generate the most revenue, but these products are also heavily discounted!

To improve revenue further, the company could try to **reduce the amount of discount offered on Adidas** products, and monitor sales volume to see if it remains stable. Alternatively, it could try **offering a small discount on Nike products** which would reduce average revenue for the brand, but may increase overall revenue if there is an increase in the volume of Nike products sold.


### Correlation between revenue and reviews
Let's explore whether relationships exist between the columns in our database.
We will check the strength and direction of the correlation between <code>revenue</code> and <code>reviews</code>.

In [115]:
%%sql

-- Calculating correlation between reviews and revenue
SELECT
    ROUND(CORR(r.reviews, f.revenue)::NUMERIC, 3) AS corr_review_revenue
FROM
    reviews AS r
  INNER JOIN
    finance AS f
  USING(product_id);

 * postgresql://postgres:***@localhost/sports
1 rows affected.


[(Decimal('0.652'),)]

There is a moderately positive correlation between <code>revenue</code> and <code>reviews</code>. This indicates that the company could focus on getting more reviews to popularize products and monitor revenue in this respect.

### Reviews by month and brand
Let's take a look at the volume of <code>reviews</code> by month to see if there are any trends or gaps we can look to exploit.

In [116]:
%%sql

-- Exploring the number of reviews in each month for both brands
SELECT
    b.brand
    , TO_CHAR(TO_TIMESTAMP(EXTRACT(MONTH FROM t.last_visited)::VARCHAR, 'MM'), 'Mon') AS month
    , COUNT(r.product_id) AS num_reviews
FROM
    brands AS b
    INNER JOIN traffic AS t USING(product_id)
    INNER JOIN reviews AS r USING(product_id)
WHERE
    b.brand IS NOT NULL
    AND EXTRACT(MONTH FROM t.last_visited) IS NOT NULL
GROUP BY
    b.brand
    , EXTRACT(MONTH FROM t.last_visited)
ORDER BY
    brand
    , EXTRACT(MONTH FROM t.last_visited);

 * postgresql://postgres:***@localhost/sports
24 rows affected.


[('Adidas', 'Jan', 253),
 ('Adidas', 'Feb', 272),
 ('Adidas', 'Mar', 269),
 ('Adidas', 'Apr', 180),
 ('Adidas', 'May', 172),
 ('Adidas', 'Jun', 159),
 ('Adidas', 'Jul', 170),
 ('Adidas', 'Aug', 189),
 ('Adidas', 'Sep', 181),
 ('Adidas', 'Oct', 192),
 ('Adidas', 'Nov', 150),
 ('Adidas', 'Dec', 190),
 ('Nike', 'Jan', 52),
 ('Nike', 'Feb', 52),
 ('Nike', 'Mar', 55),
 ('Nike', 'Apr', 42),
 ('Nike', 'May', 41),
 ('Nike', 'Jun', 43),
 ('Nike', 'Jul', 37),
 ('Nike', 'Aug', 29),
 ('Nike', 'Sep', 28),
 ('Nike', 'Oct', 47),
 ('Nike', 'Nov', 38),
 ('Nike', 'Dec', 35)]

Looks like product reviews are highest in the first quarter of the calendar year, so there is scope to run experiments aiming to increase the number of reviews in the remaining three quarters!

### Ratings and reviews by product description length
The length of a product's <code>description</code> might influence a product's <code>rating</code> and if so, the company can produce content guidelines for listing products on their website and test if this influences <code>revenue</code>.


In [117]:
%%sql

-- Average rating for different ranges of description length
SELECT
    TRUNC(LENGTH(i.description), -2) AS description_length
    , ROUND(AVG(r.rating::NUMERIC), 2) AS average_rating
FROM
    info AS i
  INNER JOIN
    reviews AS r
  USING(product_id)
WHERE
    i.description IS NOT NULL
GROUP BY
    description_length
ORDER BY
    description_length;

 * postgresql://postgres:***@localhost/sports
7 rows affected.


[(Decimal('0'), Decimal('1.87')),
 (Decimal('100'), Decimal('3.21')),
 (Decimal('200'), Decimal('3.27')),
 (Decimal('300'), Decimal('3.29')),
 (Decimal('400'), Decimal('3.32')),
 (Decimal('500'), Decimal('3.12')),
 (Decimal('600'), Decimal('3.65'))]

Unfortunately, there doesn't appear to be a clear pattern between the length of a product's <code>description</code> and its <code>rating</code>.

### Footwear product performance
So far, we have been primarily analyzing Adidas vs Nike products. Now, let's analyze the type of products being sold.

As we don't have any data on product type, we will filter <code>description</code> for keywords, to find out how much of the company's product range consists of footwear and the median <code>revenue</code> generated by these.

In [118]:
%%sql

-- Exploring the footwear product type
WITH footwear AS (
    SELECT
        i.description
        , f.revenue
    FROM
        info AS i
        INNER JOIN finance AS f USING(product_id)
    WHERE
        i.description ILIKE ANY (array['%shoe%', '%foot%'])
        AND i.description IS NOT NULL
)

SELECT
    COUNT(description) AS num_footwear_products
    , (SELECT COUNT(description) FROM info) AS total_products
    , PERCENTILE_DISC(0.5) WITHIN GROUP(ORDER BY revenue) AS median_footwear_revenue
    , (SELECT PERCENTILE_DISC(0.5) WITHIN GROUP(ORDER BY revenue) FROM finance) AS median_revenue_all
FROM
    footwear;

 * postgresql://postgres:***@localhost/sports
1 rows affected.


[(2671, 3117, Decimal('3109.43'), Decimal('2735.28'))]

Out of the total of 3,117 products, 2,671 are footwear products. **Footwear accounts for around 85%** of the company's products. They also generate a median revenue of over $3100 dollars as compared to $2700 for all the products.

### Clothing product performance

Let's examine how this compares to products other than footwear i.e. Clothing. We will count the number of products and <code>median_revenue</code> of products that are *not in <code>footwear</code>*.

In [119]:
%%sql

WITH footwear AS (
    SELECT
        i.description
        , f.revenue
    FROM
        info AS i
        INNER JOIN finance AS f USING(product_id)
    WHERE
        i.description ILIKE ANY (array['%shoe%', '%foot%'])
        AND i.description IS NOT NULL
)

SELECT
    COUNT(i.product_id) AS num_clothing_products
    , PERCENTILE_DISC(0.5) WITHIN GROUP(ORDER BY f.revenue) AS median_clothing_revenue
FROM
    info AS i
    INNER JOIN finance AS f USING(product_id)
WHERE
    i.description NOT IN (SELECT description FROM footwear);

 * postgresql://postgres:***@localhost/sports
1 rows affected.


[(446, Decimal('604.58'))]